datumaro/plugins/splitter.py

```diff
@@ -0,0 +1,480 @@
+                        break
+                    choice = np.random.choice(range(len(diffs[nearest])))
+                    pid_test, pid_trval = diffs[nearest][choice]
+                    testset_total += nearest
+                    new_diffs = dict()
+                    for diff, person_ids in diffs.items():
+                        new_list = []
+                        for id1, id2 in person_ids:
+                            if id1 == pid_test or id2 == pid_trval:
+                                continue
+                            new_list.append((id1,id2))
+                        if len(new_list)>0:
+                            new_diffs[diff] = new_list
+                    diffs = new_diffs
+                    exchanges.append((pid_test, pid_trval))
+                # exchange
+                for pid_test, pid_trval in exchanges:
+                    testset[pid_trval] = trainval.pop(pid_trval)
+                    trainval[pid_test] = testset.pop(pid_test)
+        else:
+            testset = dict()
+            trainval = by_pid
+
+        by_splits = dict()
+        for subset in self._subsets:
+            by_splits[subset] = []
+
+        ## 2. split 'test' into 'gallery' and 'query'
+        if len(testset)>0:
+            for person_id, items in testset.items():
+                indice = [idx for idx, _ in items]
+                by_splits['test'].extend(indice)
+
+            valid = ["gallery", "query"]
+            test_subsets, test_ratio = self._validate_splits(test_splits,valid)
+            required = self._get_required(test_ratio)
+            for person_id, items in testset.items():
+                np.random.shuffle(items)
+                by_attributes = self._group_by_attributes(items)
+                for attributes, indice in by_attributes.items():
+                    gname = 'person_id: {}, attributes: {}'.format(
+                        person_id, attributes)
+                    splits = self._split_indice(indice, gname,
+                        test_ratio, required)
+
+                    # tag using group
+                    for idx, item in enumerate(self._extractor):
+                        for subset, split in zip(test_subsets, splits):
+                            if idx in split:
+                                group_id = self._group_map[subset]
+                                item.annotations[0].group = group_id
+                                break
+
+        ## 3. split 'trainval' into  'train' and 'val'
+        trainval_subsets = ["train", "val"]
+        trainval_ratio = []
+        for subset in trainval_subsets:
+            if subset in id_subsets:
+                val = id_ratio[id_subsets.index(subset)]
+            else:
+                val = 0.0
+            trainval_ratio.append(val)
+        trainval_ratio = np.array(trainval_ratio)
+        total_ratio = np.sum(trainval_ratio)
+        if total_ratio < 1e-7:
```

Comment:
This number is used multiple times, maybe it's worth to be extracted as a constant.

Reply:
Thank you. I'll make it to a constant.